### ISSUE TITLE
Promise Caching Edge Case

### ISSUE THREAD

**Issue by noahrepublic** [noahrepublic/DataKeep#21]
Thank you to Nezuo for alerting me the discovery in his own module.

I will quote what he wrote:

> `Collection:load` caches load promises for each document. The promise is removed from the cache when `Document:close` is called.
> 
>1. Player joins the server and `Document:load` is called which returns a promise that is cached.
> 
>2. Player leaves and rejoins that same server.
> 
>3. `Document:load` is called again, the promise from the first load never finished (meaning `Document:close` hasn't been called yet), so the cached promise from the first load is returned.
> 
>4. The promise resolves, the original load calls `Document:close` because that `Player` instance is no longer in the game.
> 
>5. The new load now has a closed document and none of their data will be saved at this point.

### The fix

Lapis removed promise caching, but I want to search for a way around it before making a change like that.

**Comment by FireBird702** [comment]
What about marking it as closing (releasing) and waiting for it to close and then loading new one or it won't fix the problem?

**Comment by noahrepublic** [comment]
I think this solution could cause another edge case, what if it never resolves? I think instead we should try to invalidate the cache

**Comment by FireBird702** [comment]
> I think this solution could cause another edge case, what if it never resolves? I think instead we should try to invalidate the cache

Then maybe add a timer to it? 🤔

```lua
local promise = Promise.try(function()
	if Keeps[id] then -- TODO: check if got rejected before returning cache
		if not Keeps[id]._releasing then
			return Keeps[id]
		end

		-- wait for keep to be released: https://github.com/noahrepublic/DataKeep/issues/21

		local timer = 60 * 2 -- 2 minutes. Long but almost impossible to happen

		repeat
			timer -= task.wait()
		until timer < 0 or Keeps[id]._released or Keeps[id] == nil -- Keeps[id]._released will almost always be first.

		-- keep released so we can load new keep
	elseif self._cachedKeepPromises[id] and self._cachedKeepPromises[id].Status ~= Promise.Status.Rejected and self._cachedKeepPromises[id].Status ~= Promise.Status.Cancelled then
		return self._cachedKeepPromises[id]
	end

	return nil
end):andThen(function(cachedKeep)
	if cachedKeep then
		return cachedKeep
	end

	return Promise.new(function(resolve, reject)
		-- rest of the :LoadKeep() code
	end)
end)
```
^ Slightly modified `:LoadKeep()` code

**Comment by noahrepublic** [comment]
I think it should be ok to instantly remove the cache and if it is being loaded again it won't be the false cache. Just unsure how to confirm it has released and if it is needed.

**Comment by FireBird702** [comment]
> I think it should be ok to instantly remove the cache and if it is being loaded again it won't be the false cache. Just unsure how to confirm it has released and if it is needed.

Is there a chance that the new on load data save could be overwritten by the old one if you instantly remove the cache and load new keep (old keep saves after new one is loaded)?

**Comment by noahrepublic** [comment]
I am not sure I understand. But from what I think I understand I don't think so..?

**Comment by FireBird702** [comment]
I have a fix ready that doesn't involve removing caching, should I make a pull request?

**Comment by noahrepublic** [comment]
#Coming back to this, maybe a timer would be beneficial, similar to how there is a dead session timer. Just feels weird. Removing promise cache is last resort imo.

**Comment by noahrepublic** [comment]
One possible fix which kind of makes sense (it's early so please confirm this makes sense) on :Release before starting the release (so before the promise) we release any caches related to the keep (except for the release cache) since whether or not release fails we no longer know if it is accurate data.

**Comment by FireBird702** [comment]
> One possible fix which kind of makes sense (it's early so please confirm this makes sense) on :Release before starting the release (so before the promise) we release any caches related to the keep (except for the release cache) since whether or not release fails we no longer know if it is accurate data.

You will still be left with the release cache that u need to know whether it finished I guess
Check my [fork](https://github.com/FireBird702/DataKeep) which may have everything fixed already

**Comment by noahrepublic** [comment]
> You will still be left with the release cache that u need to know whether it finished I guess

Wouldn't it not matter because the release cache won't mess with the loading?

**Comment by noahrepublic** [comment]
> Check my [fork](https://github.com/FireBird702/DataKeep) which may have everything fixed already

Have you already ran tests on this fork before I begin reviewing it deeper looks really promising though

**Comment by FireBird702** [comment]
> Have you already ran tests on this fork before I begin reviewing it deeper looks really promising though

Tested new things and it looks like they are working but I'm not sure about the already existing things (mainly `:Overwrite()` and GlobalUpdates). I'll try to test that too

**Comment by noahrepublic** [comment]
I don't have the time currently to read the whole thing, what solution did you use?

**Comment by FireBird702** [comment]
I'm awaiting for the keep to finish releasing with a timeout

**Comment by noahrepublic** [comment]
How did you determine the timeout amount?

**Comment by FireBird702** [comment]
```luau
local timer = Store._assumeDeadLock -- in normal conditions there is no way to hit that but I am not sure if it should be that long anyway

repeat
	timer -= task.wait()
until (Keeps[id]._released and not Keeps[id]) or timer < 0
```

**Comment by noahrepublic** [comment]
> ```luau
> local timer = Store._assumeDeadLock -- in normal conditions there is no way to hit that but I am not sure if it should be that long anyway
> 
> repeat
> 	timer -= task.wait()
> until (Keeps[id]._released and not Keeps[id]) or timer < 0
> ```

If someone instantly rejoins wouldn't it still happen then since it isn't considered invalid yet? Unless I am misunderstanding

**Comment by FireBird702** [comment]
Oops, there was a little misunderstanding on my side

Part of the `:LoadKeep()` code:
```luau
local promise = Promise.try(function()
	if Keeps[id] then
		if not Keeps[id]._releasing and not Keeps[id]._released then
			return Keeps[id]
		end

		-- wait for keep to be released on the same server: https://github.com/noahrepublic/DataKeep/issues/21

		local timer = Store._assumeDeadLock -- in normal conditions there is no way to hit that

		repeat
			timer -= task.wait()
		until (Keeps[id]._released and not Keeps[id]) or timer < 0

		if Keeps[id] then
			releaseKeepInternally(Keeps[id]) -- additional cleanup to prevent memory leaks
		end
	elseif self._cachedKeepPromises[id] and self._cachedKeepPromises[id].Status ~= Promise.Status.Rejected and self._cachedKeepPromises[id].Status ~= Promise.Status.Cancelled then
		-- already loading keep
		return self._cachedKeepPromises[id]
	end

	-- keep released so we can load new keep

	return nil
end):andThen(function(cachedKeep)
	if cachedKeep then
		return cachedKeep
	end
		
	-- rest of the :LoadKeep() code here
end)
```

When `:Release()` is called `keep._releasing` is set to `true` at the beginning

**Comment by noahrepublic** [comment]
Ah, I see. Instead of a continuous loop couldn't you use `:timeout()` and check if it succeeded or rejected and continue depending on the result?

**Comment by FireBird702** [comment]
I'm not sure how I can change it to `:timeout()` and even after that I would still need to do a loop?

**Comment by noahrepublic** [comment]
I'm on mobile so little weird to type, but we have two options, but really only the second one makes sense, but may be able to take something from the first

First:
```luau

local promise = Promise.try(function()
    if Keeps[id] then
        if not Keeps[id]._releasing and not Keeps[id]._released then
			return Keeps[id] -- may need to use resolve()..?
		end
		
		repeat
		
		until it is released
	elseif self._cachedKeepPromises[id] and self._cachedKeepPromises[id].Status ~= Promise.Status.Rejected and self._cachedKeepPromises[id].Status ~= Promise.Status.Cancelled then
		-- already loading keep
		return self._cachedKeepPromises[id]
    end
    
    return nil
end):tap(function()
	if Keeps[id] then
		releaseKeepInternally(Keeps[id]) -- additional cleanup to prevent memory leaks
	end
end):andThen(function(cachedKeep)
	-- Continue load keep code
end)
```

This change isn't super helpful except the use of `:tap()`

OR we can avoid the loop all together by exposing `releaseCache` from Keep.lua

Second: 

```luau

local promise = Promise.all({
	Keep._releaseCache[id]:timeout(Store._assumeDeadLock),
	Promise.try(function()
		if Keeps[id] then
        if not Keeps[id]._releasing and not Keeps[id]._released then
			return Keeps[id] -- may need to use resolve()..?
		end
	elseif self._cachedKeepPromises[id] and self._cachedKeepPromises[id].Status ~= Promise.Status.Rejected and self._cachedKeepPromises[id].Status ~= Promise.Status.Cancelled then
		-- already loading keep
		return self._cachedKeepPromises[id]
    end
    
    return nil
	end)
}):tap(function()
	if Keeps[id] then
		releaseKeepInternally(Keeps[id]
	end
end):andThen(function(cachedKeep)
	-- continue load keep code
end)
```

I think this is an interesting approach, this was written on mobile so was unable to test nor have proper linting further then Working Copy provided. Let me know your thoughts

**Comment by noahrepublic** [comment]
I don't think `:andThen` is right here in the case it times out, but the idea is there. Probably would have to wrap in a new Promise in order to ensure `:tap()` and `:andThen()` runs

**Comment by FireBird702** [comment]
Second option may have sense but then `_releaseCache` still needs to know when keep is removed internally (cleanup signal connections, etc.). I'm playing with this but it seems impossible without changing how internal cleanup works, unless there is other way to know when keep is cleaned up internally (`Keeps[id] == nil`) without any sort of loop

**Comment by noahrepublic** [comment]
Is it possible to listen to the `Saving` signal? We can look for that particular keep and maybe can assume it has released cuz it saved? Otherwise not sure. But this would be really nice to be able to do.

**Comment by FireBird702** [comment]
> Is it possible to listen to the `Saving` signal? 

Tried using `.Releasing:Wait()`, but there is a chance that it already fired before and `.Releasing:Wait()` will wait infinitely

**Comment by noahrepublic** [comment]
The problem with `Releasing` is it fires **before** internal release.
